pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net/url"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/mixedversion"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+)
+
+func registerC2CMixedVersions(r registry.Registry) {
+
+	sp := replicationSpec{
+		srcNodes: 4,
+		dstNodes: 4,
+		// The timeout field ensures the c2c roachtest driver behaves properly.
+		timeout:                   10 * time.Minute,
+		workload:                  replicateKV{readPercent: 0, debugRunDuration: 1 * time.Minute, maxBlockBytes: 1, initWithSplitAndScatter: true},
+		additionalDuration:        0 * time.Minute,
+		cutover:                   30 * time.Second,
+		skipNodeDistributionCheck: true,
+		suites:                    registry.Suites(registry.Nightly),
+	}
+
+	r.Add(registry.TestSpec{
+		Name:             "c2c/mixed-versions",
+		Owner:            registry.OwnerDisasterRecovery,
+		Cluster:          r.MakeClusterSpec(sp.dstNodes + sp.srcNodes + 1),
+		CompatibleClouds: sp.clouds,
+		Suites:           registry.Suites(registry.Nightly),
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runC2CMixedVersions(ctx, t, c, sp)
+		},
+	})
+}
+
+// TODO (msbutler): schedule upgrades during initial scan and cutover.
+func runC2CMixedVersions(ctx context.Context, t test.Test, c cluster.Cluster, sp replicationSpec) {
+
+	sourceMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(1, sp.srcNodes),
+		mixedversion.MinimumSupportedVersion("v23.2.0"),
```

Comment:
ideally, i test 23.2 -> 24.1 - > 24.2 -> master for now, and i think these configs allow that.

Reply:
Yes, it does. But since the two `mixedversion.Test` instances are disconnected, they might go through different upgrade paths (i.e., `sourceMvt` could start at 24.1, and `destMvt` could start at `23.2`).

In order to not over-complicate this test, I suggest we start with a fixed number of upgrades (`mvt.NumUpgrades(2)` for example).

Reply:
Starting a single major version upgrade for now, with no version skipping.